### ISSUE TITLE
[Question] Is there a way to capture screenshot at the end of each test.step?

### ISSUE THREAD

**Issue by s-sivasubramaniam** [microsoft/playwright#20068]
I understand playwright captures screenshot at the end of each test if we keep "screenshot: on" in playwright.config.js. But is there a way to do it at the end of each test.step as well and include it in html reporter? as my project requires to capture screenshot after each step. I know it can be achieved with testInfo.attach which means i have to include the piece of code "await testInfo.attach('screenshot', { body: await page.screenshot(), contentType: 'image/png' });" in each test.step and but is there a simpler way to do it?

**Comment by aslushnikov** [comment]
@s-sivasubramaniam no, there's no way for now. Let me marked this as a feature request. If it becomes popular, we'll prioritize it accordingly!

**Comment by moshikd** [comment]
this will be a great feature. can it be implemented soon?

**Comment by aslushnikov** [comment]
@moshikd please upvote the issue to bump it's priority. The more upvotes something has, the more we look onto it!

**Comment by cheshi-mantu** [comment]
many PW / allure report users are asking for this feature gents

**Comment by vigneshwarselvaraj92** [comment]
this will be great feature

**Comment by Ash2803** [comment]
> gents

++++ waiting for this feature

**Comment by soofianazneen** [comment]
++

**Comment by aidenstern** [comment]
Would really love this feature, have a ton of "testInfo.attach"on each test.step

**Comment by etwood24** [comment]
Yes please - would be great to have this as a feature without needing testInfo.attach scatterd throughout the code.

**Comment by jonasclaes** [comment]
If this existed in the code, I would use it immediately!

I think something like a `test.beforeEachStep` and a `test.afterEachStep` function would be really useful. This would allow for other things to be called as well.

**Comment by ramsar7002** [comment]
This will be a great feature

**Comment by nizzel55** [comment]
Any news about this feature?

**Comment by vitalets** [comment]
Until this is implemented I use the following custom fixture `step`:
```ts
import { test as base, expect } from '@playwright/test';

type StepFn = (title: string, fn: () => unknown) => unknown;

const test = base.extend<{step: StepFn}>({
  step: async ({ page }, use, testInfo) => {
    await use(async (title, fn) => {
      await test.step(title, async () => {
        try {
          await fn();
        } finally {
          await testInfo.attach(`screenshot after step "${title}"`, { 
            body: await page.screenshot(), 
            contentType: 'image/png' 
          }); 
        }
      });
    });
  }
});
```
Usage in test:
```ts
  test('Check title', async ({ page, step }) => {
    await step('I open url', async () => {
       await page.goto('https://playwright.dev');
    });
    await step('I click link', async () => {
      await page.getByRole('link', { name: 'Get started' }).click();
    });
    await step('I check title', async () => {
      await expect(page).toHaveTitle('Installation | Playwright');
    })
  });
```
<details>
  <summary>Steps in HTML report:</summary>

![image](https://github.com/microsoft/playwright/assets/1473072/5a9decd6-8fb6-4f68-a9f3-21a684664d66)
</details>

<details>
  <summary>Screenshots in HTML report:</summary>

![image](https://github.com/microsoft/playwright/assets/1473072/e04fdfe7-fd7e-45ed-bc78-f7eac7799f60)
</details>

**Comment by rakhmatullov** [comment]
+ for the test step screenshotting feature to be added so that screenshots can be available in the report after each test step.

**Comment by ardikun1412** [comment]
+++

**Comment by abautu** [comment]
+1

**Comment by rykerwilliams** [comment]
+1

**Comment by wojtekbi** [comment]
still not implemented?

**Comment by manuel-sanchez-vilt** [comment]
+1

**Comment by degrootsam** [comment]
+1

**Comment by aarjithn** [comment]
It would be great to have ability to 'extend' a step like we have test.extend / test.afterEach / beforeEach - screenshot is one use case, others could be updating reporting sites like reportportal / browserstack.

https://github.com/microsoft/playwright/issues/20068#issuecomment-1924295711 is the closest, but requires bypassing test.step usage and hence leaks abstraction.
